packages/medusa/src/services/draft-order.ts

```diff
@@ -0,0 +1,430 @@
+              .withTransaction(transactionManager)
+              .applyDiscount(rawCart as Cart, code)
+          }
+        }
+
+        const createdCart = await this.cartService_
+          .withTransaction(transactionManager)
+          .create({ type: CartType.DRAFT_ORDER, ...rawCart })
+
+        const draftOrder = draftOrderRepo.create({
+          cart_id: createdCart.id,
+          no_notification_order,
+        })
+        const result = await draftOrderRepo.save(draftOrder)
+
+        await this.eventBus_
+          .withTransaction(transactionManager)
+          .emit(DraftOrderService.Events.CREATED, {
+            id: result.id,
+          })
+
+        for (const item of items) {
+          if (item.variant_id) {
+            const line = await this.lineItemService_
+              .withTransaction(transactionManager)
+              .generate(item.variant_id, data.region_id, item.quantity, {
+                metadata: item?.metadata || {},
+                unit_price: item.unit_price,
+                cart: createdCart,
+              })
+
+            await this.lineItemService_
+              .withTransaction(transactionManager)
+              .create({
+                ...line,
+                cart_id: createdCart.id,
+              })
+          } else {
+            let price
+            if (typeof item.unit_price === `undefined` || item.unit_price < 0) {
+              price = 0
+            } else {
+              price = item.unit_price
+            }
+
+            // custom line items can be added to a draft order
+            await this.lineItemService_
+              .withTransaction(transactionManager)
+              .create({
+                cart_id: createdCart.id,
+                has_shipping: true,
+                title: item.title || "Custom item",
+                allow_discounts: false,
+                unit_price: price,
+                quantity: item.quantity,
+              })
+          }
+        }
+
+        for (const method of shipping_methods) {
+          await this.cartService_
+            .withTransaction(transactionManager)
+            .addShippingMethod(createdCart.id, method.option_id, method.data)
+        }
+
+        return result
+      }
+    )
+  }
+
+  /**
+   * Registers a draft order as completed, when an order has been completed.
+   * @param draftOrderId - id of draft order to complete
+   * @param orderId - id of order completed from draft order cart
+   * @return the created order
+   */
+  async registerCartCompletion(
+    draftOrderId: string,
+    orderId: string
+  ): Promise<UpdateResult> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        return await transactionManager.update(
+          DraftOrder,
+          {
+            id: draftOrderId,
+          },
+          {
+            status: DraftOrderStatus.COMPLETED,
+            completed_at: new Date(),
+            order_id: orderId,
+          }
+        )
```

Comment:
suggestion(non-blocking): Instead of using the transaction manager directly, you could use the `DraftOrderRepository` to keep consistency in how we use the ORM.